File: 113. Path Sum II/113.go
package leetcode

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func pathSum(root *TreeNode, sum int) [][]int {
	res := make([][]int, 0)
	tmp := make([]int, 0)

	dfs(&res, tmp, root, sum)

	return res
}

func dfs(res *[][]int, tmp []int, root *TreeNode, sum int){
	if root == nil{
		return
	}

	sum -= root.Val
	tmp = append(tmp, root.Val)
	if sum==0 && root.Left==nil && root.Right==nil{
		*res = append(*res, tmp)
	}

	tmp1 := make([]int, len(tmp))
	tmp2 := make([]int, len(tmp))
	copy(tmp1, tmp)
	copy(tmp2, tmp)

	dfs(res, tmp1, root.Left, sum)
	dfs(res, tmp2, root.Right, sum)
}

File: 113. Path Sum II/_113.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        res = list()
        tmp = list()

        self.dfs(res, tmp, sum, root)
        return res

    def dfs(self, res, tmp, sum, root):
        if not root:
            return

        sum -= root.val

        if sum == 0 and not root.left and not root.right:
            res.append(tmp+[root.val])
            return
        
        self.dfs(res, tmp+[root.val], sum, root.left)
        self.dfs(res, tmp+[root.val], sum, root.right)